File: prototype/rails/src/test/domain/auth/local_redirect_test.rb
require "test_helper"

module Domain
  module Auth
    class LocalRedirectTest < ActiveSupport::TestCase
      ORIGIN = "http://localhost:3000".freeze
      FALLBACK = "/account".freeze

      def test_a_missing_target_falls_back
        assert_equal FALLBACK, resolve(nil)
      end

      def test_a_blank_target_falls_back
        assert_equal FALLBACK, resolve("   ")
      end

      def test_a_root_relative_path_is_kept
        assert_equal "/checkout?step=2", resolve("/checkout?step=2")
      end

      def test_an_absolute_url_on_this_origin_is_kept
        assert_equal "#{ORIGIN}/checkout", resolve("#{ORIGIN}/checkout")
      end

      def test_the_origin_itself_is_kept
        assert_equal ORIGIN, resolve(ORIGIN)
      end

      def test_another_host_falls_back
        assert_equal FALLBACK, resolve("http://evil.example/steal")
      end

      def test_a_host_that_only_prefixes_this_origin_falls_back
        assert_equal FALLBACK, resolve("#{ORIGIN}.evil.example/steal")
      end

      def test_a_protocol_relative_url_falls_back
        assert_equal FALLBACK, resolve("//evil.example/steal")
      end

      def test_a_backslash_escaped_path_falls_back
        assert_equal FALLBACK, resolve("/\\evil.example/steal")
      end

      def test_a_target_carrying_a_newline_falls_back
        assert_equal FALLBACK, resolve("/checkout\nSet-Cookie: x=1")
      end

      def test_keep_if_local_returns_a_local_target
        assert_equal "/checkout", LocalRedirect.keep_if_local("/checkout", origin: ORIGIN)
      end

      def test_keep_if_local_drops_a_foreign_target
        assert_nil LocalRedirect.keep_if_local("http://evil.example/steal", origin: ORIGIN)
      end

      private

      def resolve(requested)
        LocalRedirect.resolve(requested, fallback: FALLBACK, origin: ORIGIN)
      end
    end
  end
end

File: prototype/php/src/app/Domain/Auth/LocalRedirectTest.php
<?php

declare(strict_types=1);

namespace App\Domain\Auth;

it('resolves a requested target', function (?string $requested, string $expected): void {
    $origin = 'http://localhost:8000';
    $fallback = '/account';

    expect(LocalRedirect::resolve($requested, ActorType::Customer, $fallback, $origin))->toBe($expected);
})->with([
    'a missing target falls back' => [null, '/account'],
    'a blank target falls back' => ['   ', '/account'],
    'a root-relative path is kept' => ['/checkout?step=2', '/checkout?step=2'],
    'an absolute url on this origin is kept' => ['http://localhost:8000/checkout', 'http://localhost:8000/checkout'],
    'the origin itself is kept' => ['http://localhost:8000', 'http://localhost:8000'],
    'another host falls back' => ['http://evil.example/steal', '/account'],
    'a host that only prefixes this origin falls back' => ['http://localhost:8000.evil.example/steal', '/account'],
    'a protocol-relative url falls back' => ['//evil.example/steal', '/account'],
    'a backslash-escaped path falls back' => ['/\\evil.example/steal', '/account'],
    'a target carrying a newline falls back' => ["/checkout\nSet-Cookie: x=1", '/account'],
    'the seller portal falls back for a customer' => ['/seller', '/account'],
    'a path inside the seller portal falls back for a customer' => ['/seller/orders/1', '/account'],
    'an absolute seller url falls back for a customer' => ['http://localhost:8000/seller/listings', '/account'],
    'a path that only prefixes the seller portal is kept' => ['/sellers-guide', '/sellers-guide'],
    'the admin site falls back for a customer' => ['/admin', '/account'],
    'a path inside the admin site falls back for a customer' => ['/admin/customers/1', '/account'],
]);

it('keeps a seller on the seller portal', function (): void {
    expect(LocalRedirect::resolve('/seller/orders/1', ActorType::Seller, '/dashboard', 'http://localhost:8000'))
        ->toBe('/seller/orders/1');
});

it('falls back for a seller sent to the admin site', function (): void {
    expect(LocalRedirect::resolve('/admin/customers/1', ActorType::Seller, '/dashboard', 'http://localhost:8000'))
        ->toBe('/dashboard');
});

it('keeps an admin on the admin site', function (): void {
    expect(LocalRedirect::resolve('/admin/customers/1', ActorType::Admin, '/admin', 'http://localhost:8000'))
        ->toBe('/admin/customers/1');
});

it('falls back for an admin sent to the seller portal', function (): void {
    expect(LocalRedirect::resolve('/seller/orders/1', ActorType::Admin, '/admin', 'http://localhost:8000'))
        ->toBe('/admin');
});

it('keeps a local target on its own', function (): void {
    expect(LocalRedirect::keepIfLocal('/checkout', 'http://localhost:8000'))->toBe('/checkout');
});

it('drops a foreign target on its own', function (): void {
    expect(LocalRedirect::keepIfLocal('http://evil.example/steal', 'http://localhost:8000'))->toBeNull();
});
